File: src/ts/ssh/messages/sshMessage.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Buffer } from 'buffer';
import { SshDataReader, SshDataWriter } from '../io/sshData';
import { SshSessionConfiguration } from '../sshSessionConfiguration';

export interface SshMessageConstructor<T extends SshMessage = SshMessage> {
	new (): T;
}

export abstract class SshMessage {
	public get messageType(): number {
		return 0;
	}

	protected rawBytes?: Buffer;

	public toBuffer(): Buffer {
		const writer = new SshDataWriter(Buffer.alloc(16));
		this.write(writer);
		return writer.toBuffer();
	}

	public read(reader: SshDataReader): void {
		this.rawBytes = reader.buffer;

		const number = reader.readByte();
		if (number !== this.messageType) {
			throw new Error(`Message type ${number} is not valid.`);
		}

		this.onRead(reader);
	}

	public write(writer: SshDataWriter): void {
		if (this.rawBytes) {
			// Piped messages are rewritten without re-serialization. This preserves any
			// unparsed extended message data. It assumes no properties of the message
			// have been modified without also updating the serialized bytes.
			writer.write(this.rawBytes);
		} else {
			writer.writeByte(this.messageType);

			this.onWrite(writer);
		}
	}

	protected onRead(reader: SshDataReader): void {
		throw new Error('Not supported.');
	}

	protected onWrite(writer: SshDataWriter): void {
		throw new Error('Not supported.');
	}

	protected validateField<T>(value: T | undefined, name: string): T {
		if (typeof value === 'undefined') {
			throw new Error(`${this.constructor.name} ${name} is required.`);
		}

		return value;
	}

	public toString() {
		return this.constructor.name;
	}

	public static readonly index = new Map<number, { new (): SshMessage }>();

	public static create(
		config: SshSessionConfiguration,
		messageType: number,
		data: Buffer,
	): SshMessage | null {
		const messageClass = config.messages.get(messageType);
		if (messageClass) {
			const message = new messageClass();
			message.read(new SshDataReader(data));
			return message;
		} else {
			return null;
		}
	}

	public convertTo<T extends SshMessage>(otherMessage: T, copy = false): T {
		const reader = new SshDataReader(copy ? Buffer.from(this.rawBytes!) : this.rawBytes!);
		otherMessage.read(reader);
		return otherMessage;
	}
}

File: src/cs/Ssh/Messages/SshMessage.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Diagnostics;
using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Ssh.Messages;

[DebuggerDisplay("{ToString(),nq}")]
public abstract class SshMessage
{
	/// <summary>
	/// Tries to create a message with the given message type code.
	/// </summary>
	/// <returns>The constructed message object, or null if not a known message type.</returns>
	public static SshMessage? TryCreate(SshSessionConfiguration config, byte messageType)
	{
		if (config == null) throw new ArgumentNullException(nameof(config));

		if (config.Messages.TryGetValue(messageType, out var type))
		{
			return (SshMessage)Activator.CreateInstance(type)!;
		}

		return null;
	}

	public abstract byte MessageType { get; }

	protected Buffer RawBytes { get; set; }

	public void Read(ref SshDataReader reader)
	{
		RawBytes = reader.Buffer;

		var number = reader.ReadByte();
		if (number != MessageType)
		{
			throw new ArgumentException($"Message type {number} is not valid.");
		}

		OnRead(ref reader);
	}

	public void Write(ref SshDataWriter writer)
	{
		if (this.RawBytes.Count > 0)
		{
			// Piped messages are rewritten without re-serialization. This preserves any
			// unparsed extended message data. It assumes no properties of the message
			// have been modified without also updating the serialized bytes.
			writer.WriteRaw(this.RawBytes);
		}
		else
		{
			writer.Write(MessageType);

			OnWrite(ref writer);
		}
	}

	public Buffer ToBuffer()
	{
		var writer = new SshDataWriter();
		Write(ref writer);
		return writer.ToBuffer();
	}

	public T ConvertTo<T>(bool copy = false) where T : SshMessage, new()
	{
		var reader = new SshDataReader(copy ? RawBytes.Copy() : RawBytes);
		var message = new T();
		message.Read(ref reader);
		return message;
	}

	protected virtual void OnRead(ref SshDataReader reader)
	{
		throw new NotSupportedException();
	}

	protected virtual void OnWrite(ref SshDataWriter writer)
	{
		throw new NotSupportedException();
	}

	public override string ToString()
	{
		return GetType().Name;
	}
}
